Sources/ConnectionPoolModule/PoolStateMachine.swift

```diff
@@ -84,10 +116,452 @@ struct PoolStateMachine<
+    private(set) var poolState: PoolState = .running
+    @usableFromInline
+    private(set) var cacheNoMoreConnectionsAllowed: Bool = false
+
+    @usableFromInline
+    private(set) var failedConsecutiveConnectionAttempts: Int = 0
+    
+    @inlinable
+    init(
+        configuration: PoolConfiguration,
+        generator: ConnectionIDGenerator,
+        timerCancellationTokenType: TimerCancellationToken.Type
+    ) {
+        self.configuration = configuration
+        self.generator = generator
+        self.connections = ConnectionGroup(
+            generator: generator,
+            minimumConcurrentConnections: configuration.minimumConnectionCount,
+            maximumConcurrentConnectionSoftLimit: configuration.maximumConnectionSoftLimit,
+            maximumConcurrentConnectionHardLimit: configuration.maximumConnectionHardLimit,
+            keepAlive: configuration.keepAliveDuration != nil,
+            keepAliveReducesAvailableStreams: true
+        )
+        self.requestQueue = RequestQueue()
+    }
+
+    mutating func refillConnections() -> [ConnectionRequest] {
+        return self.connections.refillConnections()
+    }
+
+    @inlinable
+    mutating func leaseConnection(_ request: Request) -> Action {
+        switch self.poolState {
+        case .running:
+            break
+
+        case .shuttingDown, .shutDown:
+            return .init(
+                request: .failRequest(request, ConnectionPoolError.poolShutdown),
+                connection: .none
+            )
+        }
+
+        if !self.requestQueue.isEmpty && self.cacheNoMoreConnectionsAllowed {
+            self.requestQueue.queue(request)
+            return .none()
+        }
+
+        var soonAvailable: UInt16 = 0
+
+        // check if any other EL has an idle connection
+        switch self.connections.leaseConnectionOrSoonAvailableConnectionCount() {
+        case .leasedConnection(let leaseResult):
+            return .init(
+                request: .leaseConnection(TinyFastSequence(element: request), leaseResult.connection),
+                connection: .cancelTimers(.init(leaseResult.timersToCancel))
+            )
+
+        case .startingCount(let count):
+            soonAvailable += count
+        }
+
+        // we tried everything. there is no connection available. now we must check, if and where we
+        // can create further connections. but first we must enqueue the new request
+
+        self.requestQueue.queue(request)
+
+        let requestAction = RequestAction.none
+
+        if soonAvailable >= self.requestQueue.count {
+            // if more connections will be soon available then we have waiters, we don't need to
+            // create further new connections.
+            return .init(
+                request: requestAction,
+                connection: .none
+            )
+        }
+
+        // Can we create a demand connection
+        if let request = self.connections.createNewDemandConnectionIfPossible() {
```

Comment:
Can this and the following `if` be `else if` clauses (and the last couple lines be an `else` clause)? It'd be much more readable to me that way.